torch/csrc/distributed/c10d/ProcessGroupGloo.cpp

```diff
@@ -879,15 +888,43 @@ void ProcessGroupGloo::runLoop(int workerIndex) {
     workConsumeCV_.notify_one();
 
     AsyncWork::execute(work);
+    // TODO: Need to find a way to calculate the difference of duration of two
+    // c10d::Event
+    FlightRecorder::get()->retire_id(work->trace_id_, false);
     lock.lock();
     workInProgress_[workerIndex].reset();
   }
 }
 
+const std::vector<uint64_t>& ProcessGroupGloo::groupRanks() const {
+  if (options_->global_ranks_in_group.empty() && local_id_ == 0) {
+    static std::vector<uint64_t> globalRanks(size_);
+    std::iota(globalRanks.begin(), globalRanks.end(), 0);
+    return globalRanks;
+  }
+  return options_->global_ranks_in_group;
+}
+
 void ProcessGroupGloo::enqueue(c10::intrusive_ptr<AsyncWork> work) {
   std::unique_lock<std::mutex> lock(workMutex_);
   workQueue_.push_back(std::move(work));
   lock.unlock();
+  // using c10d::FlightRecorder;
+  work->trace_id_ = FlightRecorder::get()->record(
```

Comment:
this part looks right to me. (straightforward..) - recording event at the time of dispatching the async kernel.

however what i'm not clear on is how gloo implements timeouts.  Does it even have an equivalent of a watchdog thread, or is it going to wait until someone checks the result of a future to finally raise the timeout exception?

Reply:
actually, is it safer to record the FR event _before_ acquiring the workMutex?

i'm not sure under what circumstances gloo could deadlock, but the question is whether we'd want to have a record in FR for the about-to-enqueue work that didn't enqueue due to a deadlock

Reply:
It is using a side thread to run the collective as well.. But we just use that thread to call exec() and IIUC lots of collectives are blocking so there is no timeout check even. By tracing the code inside gloo, we do the waitSend/waitRecv with timeout inside the implementation of gloo. So there is no way we can do the timeout check in c10d layer, at least for now.

And for recording, I prefer to record it when everything is settled. Because since gloo is synchronous we know which collective will timeout exactly.